Add "Software bridge management" design doc

ykulazhenkov opened the pull request:
This document contains a proposal to add limited support for software bridges configuration to the **sriov-network-operator**.

This feature assumes integration with [ovs-cni](https://github.com/k8snetworkplumbingwg/ovs-cni) and [accelerated-bridge-cni](https://github.com/k8snetworkplumbingwg/accelerated-bridge-cni).

Depends on [_switchdev and systemd modes refactoring_](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/doc/design/switchdev-refactoring.md) feature

cc @adrianchiris @zeeke @SchSeba

wizhaoredhat commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+`accelerated-bridge-cni` and contains the required resource requests.
+* As a user, I want to define configuration for software bridges inside the `SriovNetworkNodePolicy` CR and expect that the
+operator will create required bridges, configure them, and attach uplinks (physical functions).
+
```
What about the OVN-K CNI use case? Will the api to set it remain the same?
Will OVN-K and OVS-CNI have any conflicts if OVN-K is installed in the first place.

ykulazhenkov replied:
I believe that the only difference for OVN-k use case is that sriov-network-operator will need to be configured to `systemd` mode (check [this doc](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/doc/design/switchdev-refactoring.md) for details).  [NIC configuration only flow](https://github.com/ykulazhenkov/sriov-network-operator/blob/3753c5dcc4cf83a46a85bf1334f6ada31d9a22ce/doc/design/software-bridge-management.md?plain=1#L110) is proposed to be compatible with existing OVN-k use-case. In this case NIC will be configured to switchdev mode early on boot by the systemd service, but the operator will not manage software bridges.

ykulazhenkov replied:
The current proposal assumes that "Fully automatic" flow can be used only for dedicated NICs (not used by OVN-k) which are fully controlled by the sriov-network-operator.

zeeke commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	ExternallyManaged bool              `json:"externallyManaged,omitempty"`
+	TotalVfs          int               `json:"totalvfs,omitempty"`
+	VFs               []VirtualFunction `json:"Vfs,omitempty"`
+	Bridge            *Bridge           `json:"bridge,omitempty"`
```
Is this the same type as we have in the `SriovNetworkNodeState.Spec` and in the `SriovNetworkNodeState.Spec`? If this information comes from an inspection on the host, it might reflect the output of a `ovs-vsctl` or `ip link`, right?

ykulazhenkov replied:
yes, currently I think that it is possible to use the same type here and in the policy/state.

> `ovs-vsctl` or `ip link`
just as a  note, I think we will use direct requests to ovsd  to configure ovs bridges and rely on netlink for linux bridges.
Usage of cmd tools is not expected.

ykulazhenkov replied:
> agree we should use netlink (if we miss something we can add to the golang netlink) and for ovsdb I think we can use the same package ovn-k use?

yes, I think we have options to avoid usage of cmd tools

zeeke commented on doc/design/software-bridge-management.md:
````diff
@@ -0,0 +1,460 @@
+```
+
+Open questions:
+- do we want to support each OVS Bridge/Interface option explicitly or should we support
````
I'd personally go with a generic container for options, describing in the docs in which command they will be used.

ykulazhenkov replied:
> do we have any option which is mandatory for the use-cases in this doc ?

No. I assume that ovs-kernel (with TC based HW offloading) should work without extra options.

User will need to:
a) enable HW offload 
b) create policy that contains Bridge.type = "ovs", without extra OVS options
c) create OVSNetwork CR

zeeke reviewed (commented):
Left a few comments. Overall design looks good to me

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+- need to clarify list of required args for OVS and Linux bridge
+
+_**Note**: multiple policies can match single PF (vf range use-case), bridge settings from the policy with the higher priority
+will be applied to PF._
```
done

adrianchiris commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	// Trunk configuration for the OVS port
+	Trunk []*OVSTrunkConfig `json:"trunk,omitempty"`
+	// OpenFlow port number in range 1 to 65,279
+	OfportRequest uint `json:"ofport_request,omitempty"`
```
do we want to have this configurable for our usecase ?

ykulazhenkov replied:
For our use-case we don't need this. But I think we should support all options from ovs-cni. WDYT?

adrianchiris commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+
+// OVSInterfaceConfig contains some options from the Interface table of the OVSDB for PF
+type OVSInterfaceConfig struct {
+	Type        string            `json:"type,omitempty"`
```
what values can be provided to type ? (dpdk / netdev ?)

ykulazhenkov replied:
yes. I think we should not filter them on the operator level. User may want to use some vendor-specific values for this field.

adrianchiris commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	// contains optional config for OVS bridge
+	Ovs *OVSConfig `json:"ovs,omitempty"`
+	// contains optional config for LinuxBr bridge
+	LinuxBr *LinuxBrConfig `json:"linux,omitempty"`
```
nit: why not name this "Linux" ?

ykulazhenkov replied:
renamed

adrianchiris commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+
+// LinuxBrConfig optional configuration for Linux bridge and uplink interface
+type LinuxBrConfig struct {
+	Interface map[string]string    // TODO clarify required settings
```
can you give some examples of this field ?

ykulazhenkov replied:
This field should hold PF's port-level bridge configuration . 
It is not clear which setting we should put here. 
VLAN configuration for the PF is a good candidate, but I'm not sure that we need it.
By default OVS doesn't require to explicitly allow VLANs on the PF interface. 
Linux bridge requires to explicitly configure VLANs on PF, but accelerated-bridge-cni already has an option to handle this.

zeeke commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+* add bridge auto-selection logic to `ovs-cni`
+* define `OVSNetwork` CRD and implement controller for it
+
+##### Phase 2
```
+1 though this is an implementation detail. 

To me, in the design doc it's enough to refer to the config-daemon component.

ykulazhenkov replied:
I removed the mention of the generic_plugin from the doc. Let's discuss this during the implementation.

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+
+    _**Note 2:** we should create udev rule which will set `NM_UNMANAGED=1` for bridges created by the operator_
+
+4. `sriov-network-config-daemon` should report information about software bridges in the status field of the `SriovNetworkNodePolicy` CR.
```
If "fully auto" mode is used  we should create a bridge with predictable name for each PF that match a policy that contains a bridge configuration. We can't use interface name to generate a predictable name for the bridge, because the limit for the interface length is only 15 symbols and PF name will already have length around this. 
I propose to use PF's PCI address to generate predictable bridge name, like `br-0000_d8_01.0`.

ykulazhenkov replied:
Yes, we will report this information in the status field of the state object.

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+
+In this case, **sriov-network-operator** handles PFs configuration, announcement of SRIOV resources with device plugin and preparation of `NetworkAttachmentDefinition` CR.
+
+In some scenarios, it may be compatible with `configurationMode: daemon`, but it is supposed to be used when the operator runs in `systemd` mode.
```
> when would it not be compatible with daemon mode ?

When `NIC configuration only flow` is used with `systemd` configurationMode, then you can apply any network configuration between  steps when VF is created  and when VF is configured(VF binds to the driver). In the `daemon` mode you don't have such option, you can only apply your custom network configuration before the operator or after, but not in the middle. This can be a problem for some configuration, e.g. for VF LAG for NVIDIA NICs

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	// OpenFlow port number in range 1 to 65,279
+	OfportRequest uint `json:"ofport_request,omitempty"`
+	// The type of interface on ovs.
+	InterfaceType string `json:"interface_type,omitempty"`
```
this is required for ovs-dpdk

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+ generic map[string]string struct which allows set any field in the Bridge/Interface tables?
+- need to clarify list of required args for OVS and Linux bridge
+
+_**Note**: multiple policies can match single PF (vf range use-case), bridge settings from the policy with
```
> what will happen when multiple PFs match the policy ?

A separate bridge will be created for each PF. I added additional note about this to the doc

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+
+#### Fully automatic workflow
+
+This workflow assumes that **sriov-network-operator** handles PFs configuration, creation and configuration of software bridge, announcement of SRIOV resources with device plugin and preparation of `NetworkAttachmentDefinition` CR.
```
done

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	IPAM string `json:"ipam,omitempty"`
+	// name of the OVS bridge, if not set OVS will automatically select bridge
+	// based on VF PCI address
+	BrName string `json:"bridge,omitempty"`
```
I renamed the field to "Bridge"

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,460 @@
+	OfportRequest uint `json:"ofport_request,omitempty"`
+	// The type of interface on ovs.
+	InterfaceType string `json:"interface_type,omitempty"`
+	// The type of interface on ovs.
```
fixed

adrianchiris reviewed (commented):
Had some discussions with @ykulazhenkov regarding possible alternatives and i believe what is currently proposed is a good way forward so overall im OK with the proposed approach.

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// The type of interface on ovs.
+	InterfaceType string `json:"interface_type,omitempty"`
+	// Path to the configuration file on the host
+	ConfigurationPath string `json:"configuration_path,omitempty"`
```
can you give an example for this?
not sure I want to expose something like that to the user

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// Paths to OVSDB socket
+	SocketFile string `json:"socket_file,omitempty"`
+	// Amount of retries for link state check
+	LinkStateCheckRetries int `json:"link_state_check_retries,omitempty"`
```
we will have a default for this?

ykulazhenkov replied:
no, ovs-cni will handle default value. If `LinkStateCheckRetries` is zero we will not add it to CNI args of the ovs-cni

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// Amount of retries for link state check
+	LinkStateCheckRetries int `json:"link_state_check_retries,omitempty"`
+	// Interval between retries for link state check
+	LinkStateCheckInterval int `json:"link_state_check_interval,omitempty"`
```
we will have a default for this?

ykulazhenkov replied:
no, ovs-cni will handle default value. If `LinkStateCheckRetries` is zero we will not add it to CNI args of the ovs-cni

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// VLAN Trunk configuration
+	Trunk []BridgeTrunkConfig `json:"trunk,omitempty"`
+	// enable setting matching vlan tags on the bridge uplink interface, default is false
+	SetUplinkVlan bool `json:"setUplinkVlan,omitempty"`
```
do we want something like this also for the ovsBridge API?

ykulazhenkov replied:
OVSNetwork and BridgeNetwork CRDs are wrappers to create CNI args for ovs-cni and accelerated-bridge-cni. 
OVS-cni doesn't have such option as for now. It is not critical to have it in the OVS-cni, because by default ovs allows all vlans on a trunk port. Linux bridge require to set vlans on a trunk port explicitly.

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+}
+
+// BridgeTrunkConfig contains configuration for OVS trunk
+type BridgeTrunkConfig struct {
```
can we use just the same TrunkConfig for both ?

ykulazhenkov replied:
yes. we ca

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+
+// OVSBridgeConfig contains some options from the Bridge table in OVSDB
+type OVSBridgeConfig struct {
+	DatapathType string            `json:"datapath_type,omitempty"`
```
Is this datapathType can be different from the one in the ovsNetwork?

ykulazhenkov replied:
yes

SchSeba replied:
can you give an example what is datapathType?
is like vdpa dpdk stuff like that?
do we want an enum with supported options?

ykulazhenkov replied:
yes: like netdev, dpdk... 
`datapath_type` can contain proprietary value, it is better not to validate it on the API level

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+}
+
+// OVSInterfaceConfig contains some options from the Interface table of the OVSDB for PF
+type OVSInterfaceConfig struct {
```
can you give some example for this field I can see everything is string and map[string]string :P

ykulazhenkov replied:
This is raw options from Interface TABLE of the OVSDB. You can check available builtin options in the OVS docs https://www.openvswitch.org/support/dist-docs/ovs-vswitchd.conf.db.5.html. This options will be set for PF interface while attaching it to OVS bridge.
These maps can contain options from upstream OVS or can be used to inject vendor specific options if required. Example for NVIDIA https://docs.nvidia.com/doca/sdk/openvswitch+offload/index.html#src-2571330903_OpenvSwitchOffload-OVS-DOCAHardwareOffloads

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+- need to clarify list of required args for OVS and Linux bridge
+
+_**Note 1**: multiple policies can match single PF (vf range use-case), bridge settings from the policy with
+ the higher priority (which is the one with the lowest `.spec.priority` value) will be applied to PF._
```
should be just tell the user that only 1 policy per PF can contain bridge section and block that via a webhook?

ykulazhenkov replied:
Not sure. I think it is better to preserve existing behavior. Currently we follow this rule for existing PF level options defined in the policy (linkType, eswitchMode, etc)

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	EswitchMode       string            `json:"eSwitchMode,omitempty"`
+	ExternallyManaged bool              `json:"externallyManaged,omitempty"`
+	TotalVfs          int               `json:"totalvfs,omitempty"`
+	VFs               []VirtualFunction `json:"Vfs,omitempty"`
```
in the virtual functions do we want to also add the representorName?

ykulazhenkov replied:
I think this is not directly related to the scope of this proposal, but I agree that this is a good improvement.

SchSeba replied:
can we add the field to represent the change in the API as part of this proposal?

ykulazhenkov replied:
ok

SchSeba commented on doc/design/software-bridge-management.md:
````diff
@@ -0,0 +1,495 @@
+}
+
+```
+`SriovNetworkNodeState.spec` and `SriovNetworkNodeState.status` should be extended to contain same `Bridge` struct as `SriovNetworkNodePolicySpec` CRD. 
````
I don't think we can use the same object as the Policy as in the policy I can have nicSelector: vendor: 8086 
but in the spec and status for the NodeState I will like to have a list of all the bridges we need to create with the configuration, name and uplink object

ykulazhenkov replied:
@SchSeba I updated proposed API change for SriovNetworkNodeState object. Now it contains Bridges list.
I propose to use the same Bridges object in the spec and in the status. We may want to extend the one in the status field with additional info when we will have feedback from the users.

SchSeba reviewed (commented):
great document!
I left some comments nice work :)

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+* add `OVSNetwork` and `BridgeNetwork` CRDs as an API for end users to simplify creation of `NetworkAttachmentDefinition` CR
+for `ovs-cni` and `accelerated-bridge-cni`
+* extend `SriovNetworkNodePolicy` CRD to support configuration of software bridges (bridge-level configuration)
+* extend `SriovNetworkNodeState` CRD to support configuration of software bridges (bridge-level configuration)
```
yes, I added a note about this

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+
+6. SRIOV resources are announced by the Device plugin
+
+7. User creates `OVSNetwork` or `BridgeNetwork` CR to create `NetworkAttachmentDefinition` CR which relies on resource announced by the Device Plugin
```
yes, good idea

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+
+4. User creates `OVSNetwork` or `BridgeNetwork` CR to create `NetworkAttachmentDefinition` CR, which relies on resources announced by the Device Plugin
+
+### API Extensions
```
Original thought was to control this with environment variable -  skip deployment if `OVS_CNI_IMAGE` is explicitly set to empty string.
WDYT?

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+// only one bridge type can be set
+type Bridge struct {
+	// +kubebuilder:validation:Enum=ovs;linux
+	Type string `json:"type"`
```
I think we can't. Structs contain optional arguments for bridges.
Minimal valid config is 
```
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetworkNodePolicy
metadata:
  name: connectx6dx-switched
  namespace: sriov-network-operator
spec:
  eSwitchMode: switchdev
  nicSelector:
    deviceID: 101d
    vendor: 15b3
  nodeSelector:
    feature.node.kubernetes.io/network-sriov.capable: "true"
  numVfs: 4
  resourceName: connectx6dx-switched
  bridge:
    type: ovs
```

In this case OVS bridge with default configuration will be created for each PF that match the policy. 
By default configuration I mean config which OVS uses if you exec `ovs-vsctl add-br foo` (no extra args).
If HW offloading is enable on the node level (with old ConfigPool API), then such bridge will work nicely with OVS-kernel hw offloading.

ykulazhenkov replied:
ok, let's drop the type field

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+}
+
+// LinuxBrConfig optional configuration for Linux bridge and uplink interface
+type LinuxBrConfig struct {
```
We have LinuxBrBridgeConfig struct which contains bridge level config for linux bridge. Then we will need to rename it to LinuxBridgeBridgeConfig which is not nice :) Let's discuss this during the implementation

SchSeba replied:
the other one can be just BridgeConfig as we are already inside the LinuxBridgeConfig section no?

ykulazhenkov replied:
ok, fixed

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+
+2. User creates `SriovNetworkNodePolicy` where:
+    * eswitch mode set to `switchdev`
+    * numVFs set to match amount of precreated VFs
```
fixed

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// based on VF PCI address
+	Bridge string `json:"bridge,omitempty"`
+	// VlanTag to assign for the OVS port
+	VlanTag uint `json:"vlan,omitempty"`
```
fixed

ykulazhenkov commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,495 @@
+	// Path to the configuration file on the host
+	ConfigurationPath string `json:"configuration_path,omitempty"`
+	// Paths to OVSDB socket
+	SocketFile string `json:"socket_file,omitempty"`
```
Not sure. My assumption was that we want to expose all user-faced options available as STDIN args in ovs-cni.  (except device-id). I think it is better to keep this option, but if it is not set explicitly by user, then we can try to use value from the daemon config.

SchSeba reviewed (commented):
I did another round of review.

general point about expose everything from the cnis args to the user.
maybe we can remove the ones I add a comment on and add them later if we see a use-case?
adding is easier than removing I think :) 

and please one critical comment if you can take a look
https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/640/files#r1518902408

ykulazhenkov commented:
thanks @SchSeba 

> general point about expose everything from the cnis args to the user. maybe we can remove the ones I add a comment on and add them later if we see a use-case? adding is easier than removing I think :)

Ok, let's remove these field for now. I removed them from the design


> and please one critical comment if you can take a look https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/640/files#r1518902408

I replied to this comment and updated the design.

PTAL

SchSeba commented on doc/design/software-bridge-management.md:
```diff
@@ -0,0 +1,508 @@
+type LinuxBridgeUPlinkConfigExt struct {
+	PciAddress string             `json:"pciAddress"`
+	Name       string             `json:"name,omitempty"`
+	Interface  OVSInterfaceConfig `json:"interface,omitempty"`
```
`OVSInterfaceConfig` -> the link bridge one right?

SchSeba commented on doc/design/software-bridge-management.md:
````diff
@@ -0,0 +1,508 @@
+}
+
+```
+`SriovNetworkNodeState.spec` and `SriovNetworkNodeState.status` should be extended to contain same `Bridges`.
````
is this statement still correct?

ykulazhenkov replied:
Yes, it is correct. The statement differs from the original one - now is says that we use the same struct in "spec" and "status" of the SriovNetworkNodeState. Also I added note which says that we may want to extend/replace the struct in the future.

SchSeba reviewed (commented):
almost ready just two small nits :)

ykulazhenkov commented:
I pushed small update to the document:

* use camelCase for json fields to follow k8s guidlines
* add featureFlag that controls state of the feature

ykulazhenkov commented:
@SchSeba @zeeke 
Could you please take another look at the document? Let's merge it if we have agreement about proposed APIs.
Thx

zeeke reviewed (approved):
LGTM
